### PR TITLE
os/bluestore: Disable NCB functionality on rotational drives

### PR DESCRIPTION
**Pull request by benhanokh**
NCB code needs to recover allocation map after an OSD crash.
The recovery process on rotational drives is about 20x slower than SSD making this solution unacceptable for that environment 

Signed-off-by: Gabriel Benhanokh <gbenhano@redhat.com>

## Checklist
- Tracker (select at least one)
  - [ ] References tracker ticket
  - [ ] Very recent bug; references commit where it was introduced
  - [x] New feature (ticket optional)
  - [ ] Doc update (no ticket needed)
  - [ ] Code cleanup (no ticket needed)
- Component impact
  - [ ] Affects [Dashboard](https://tracker.ceph.com/projects/dashboard/issues/new), opened tracker ticket
  - [ ] Affects [Orchestrator](https://tracker.ceph.com/projects/orchestrator/issues/new), opened tracker ticket
  - [x] No impact that needs to be tracked
- Documentation (select at least one)
  - [ ] Updates relevant documentation
  - [x] No doc update is appropriate
- Tests (select at least one)
  - [ ] Includes [unit test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/tests-unit-tests/)
  - [ ] Includes [integration test(s)](https://docs.ceph.com/en/latest/dev/developer_guide/testing_integration_tests/)
  - [ ] Includes bug reproducer
  - [x] No tests

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard cephadm`
- `jenkins test api`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`
- `jenkins test windows`
</details>

### CODE REVIEW

File: src/os/bluestore/BlueStore.cc [hunk 1]
```diff
@@ -6384,7 +6384,7 @@ int BlueStore::_open_db_and_around(bool read_only, bool to_repair)
   }
 
   // when function is called in repair mode (to_repair=true) we skip db->open()/create()
-  if (!read_only && !to_repair && cct->_conf->bluestore_allocation_from_file
+  if (!is_rotational() && !read_only && !to_repair && cct->_conf->bluestore_allocation_from_file
```
Comment (jdurgin) [inline src/os/bluestore/BlueStore.cc:6387]: seems like we want a new function BlueFS::is_db_rotational() -- just like the existing BlueFS::is_wal_rotational() -- which would check the db bdev, or if none, the block bdev in the same way.

BlueStore::is_rotational() is checking the main data block device afaict

**Comment by ifed01** [comment]
Just a general comment - IMO disabling NCB for spinners is a wrong direction  - it's rather hiding the NCB stuff inefficiency. 
Let's assume HDD throughput is about 100MB/s. How large allocation map could be? For 20TB hard disk and 4K min alloc size and a bit per allocation unit it would be - 20*2^40 / 2^12 / 2^3 = 20 * 2^25 = 671MB. I.e. allocation map reading from disk in the optimal case should take ~7secs. For SSD (with 500MB/s BW) this would take less then 2 secs. I.e. hdd drive specific overhead is just 5 secs. All the other stuff is pretty the same for both HDD and SSD.

The above assessment implicitly relies on hdd bandwidth for sequential access. Random one might change the picture dramatically. 
But may be we can just preallocate that contiguous space at the target device on e.g. mkfs and benefit from that?

**Comment by benhanokh** [comment]
Igor,
Restoring allocation map is faster for spinning-drive than reading Column-B from RocksDB.
The problem is recovery time when OSD crashes and no allocation map exists.
In this case we are iterating over all ONodes in RocksDB which is much slower on spinners than SSD.
As a side note, I suspect that the customer reporting the slow recovery time had very fragmented RocksDB as the sequential nature of RocksDB files should have resulted with better recovery time

**Comment by ifed01** [comment]
> Igor, Restoring allocation map is faster for spinning-drive than reading Column-B from RocksDB. The problem is recovery time when OSD crashes and no allocation map exists. In this case we are iterating over all ONodes in RocksDB which is much slower on spinners than SSD. As a side note, I suspect that the customer reporting the slow recovery time had very fragmented RocksDB as the sequential nature of RocksDB files should have resulted with better recovery time

Ah, OK... I missed that this is more about recovery from DB not the disk...

**Comment by benhanokh** [comment]
I still think we should be able to restore much faster from a spinning driver than the time reported by the customer as RocksDB SST files are sequential.

**Review by jdurgin — APPROVED** [review]
Implementation looks good, I'd suggest moving is_db_rotational() to BlueStore instead of ObjectStore since we're only using it within the store, and 'db' isn't really a concept that applies to other ObjectStores the same way

**Comment by benhanokh** [comment]
jenkins test make check (arm64)

**Comment by benhanokh** [comment]
jenkins test make check

**Comment by benhanokh** [comment]
make check

**Comment by jdurgin** [comment]
jenkins test make check

**Comment by neha-ojha** [comment]
jenkins test make check

**Review by aclamk — APPROVED** [review]

**Comment by benhanokh** [comment]
test make check

**Comment by benhanokh** [comment]
jenkins make check

**Comment by benhanokh** [comment]
jenkins retest this please

**Comment by benhanokh** [comment]
jenkins retest docs

**Comment by benhanokh** [comment]
jenkins docs

**Comment by benhanokh** [comment]
jenkins retest this please

**Pull request merged**